staging/vhost-device-sound/src/audio_backends/pipewire.rs

```diff
@@ -331,54 +344,106 @@ impl AudioBackend for PwBackend {
-                            let streams = streams
-                                .get_mut(stream_id as usize)
-                                .expect("Stream does not exist");
-                            let Some(buffer) = streams.buffers.front_mut() else {
-                                return;
-                            };
-
-                            let mut start = buffer.pos;
-
-                            let avail = (buffer.desc_len() - start as u32) as i32;
-
-                            if avail < n_bytes as i32 {
-                                n_bytes = avail.try_into().unwrap();
+                        match spa_dir {
+                            spa::Direction::Input => {
+                                let datas = buf.datas_mut();
+                                let data = &mut datas[0];
+                                let mut n_samples = data.chunk().size() as usize;
+                                if let Some(slice) = data.data() {
+                                    let mut streams = streams.write().unwrap();
+                                    let streams = streams
+                                        .get_mut(stream_id as usize)
+                                        .expect("Stream does not exist");
+
+                                    let mut start = 0;
+                                    while n_samples > 0 {
+                                        let Some(buffer) = streams.buffers.front_mut() else {
+                                            return;
+                                        };
+
+                                        let mut buf_pos = buffer.pos;
+                                        let avail = (buffer.desc_len() as usize - buf_pos) as i32;
+                                        let n_bytes = n_samples.min(avail.try_into().unwrap());
+                                        let p = &slice[start..start + n_bytes];
+
+                                        if buffer
+                                            .write_input(p)
+                                            .expect("Could not write data to guest memory")
+                                            == 0
+                                        {
+                                            break;
+                                        }
+
+                                        buf_pos += n_bytes;
+                                        buffer.pos = buf_pos;
+                                        n_samples -= n_bytes;
+                                        start += n_bytes;
+
+                                        if buf_pos >= buffer.desc_len() as usize {
+                                            streams.buffers.pop_front();
+                                        }
+                                    }
+                                    n_samples
+                                } else {
+                                    0
+                                };
```

Comment:
Here

Reply:
using return here instead cargo clippy says is an unneeded statement.
```
warning: unneeded `return` statement
   --> vhost-device-sound/src/audio_backends/pipewire.rs:385:41
    |
385 |   ...                   } else {
    |  _______________________________^
386 | | ...                       return;
    | |________________________________^

```

Reply:
Yes by "Then the else { 0 }; later is unnecessary." I meant to say you should remove the else { } branch altogether! :)

Reply:
```
                                } else {
                                    0
                                };
---------------------------------^
```

The semicolon at the end seems to make clippy not complain about the expression returning an integer because it turns it into a statement.

Reply:
ok fixed.